### DISCUSSION TITLE
In version 1.3  some bar Multi-Track Grace Notes not aligning right beside note

### DISCUSSION THREAD

**Discussion by goodgame365** [CoderLine/alphaTab#1092] [Q&A] [answered]
the erro happen When  Show two tracks at the same time  Starting from  Bar  31

Some are right, some are not，Cause unknown

the erro like this img

![未命名-1](https://user-images.githubusercontent.com/17884664/211943999-edd6a596-5bdb-4fe7-b4be-6462f8803f05.png)

it's  gp file

[test2.zip](https://github.com/CoderLine/alphaTab/files/10396903/test2.zip)

**Comment by goodgame365** [comment]
I found Cancel the following line of code and it will be normal，But it may cause new bugs.

src\rendering\glyphs\VoiceContainerGlyph.ts
------------------------------------
55                        //  let graceTargetPreBeat = this.renderer.layoutingInfo.springs.get(graceDisplayStart)!.preBeatWidth;
56                        // move right in front to the note
57                       // currentBeatGlyph.x -= graceTargetPreBeat;
-----------------------------------

![22323](https://user-images.githubusercontent.com/17884664/211991989-1ffe6a64-111b-476a-b039-48f8d4eec496.png)

**Comment by Danielku15** [comment]
I would say currently it is a mix of "by-design" from the layout algorithm and a bit a left-over from previous grace note positioning.

Generally the glyphs before note heads are aligned to not have overlaps because certain glyphs are expected to be at the same horizontal place when they are on all staves. As you indicate: just moving things left/right will rather cause other bugs. 

I agree that the positioning is not looking good, but it is also not necessarily wrong. 

Can you provide your test file so I can add an issue for improving the positioning logic?

Generally I think it should be not a big deal to position grace notes "as close to the note as possible" but do avoid bugs quite some potential offsets and glyph sizes have to be considered to avoid overlaps in case they exist. 

The technical details behind the algorithm are: 
For each time position in a bar where notes are placed, there is a "slot" created which reserves space for the glyphs before the note (pre notes) the note head itself (on note) and after the note (post note). This way we ensure that the glyphs are properly aligned in a timely fashion and respect the required size needed. The brush stroke (arrow) reserves some pixels space on the mentioned places and this width is reserved for all staves. If the second staff would also have an pre beat glyphs, this space would be first filled. 

Grace notes always have been a bit of a pain in positioning due to their timely nature of stealing their play time from the previous beat or the next beat. In one of the previous releases we changed the placement of grace notes to "stick" to the next note always and reserve space accodingly.

A proper solution/improvement for your reported problem would need: Remembering the actually needed space on the staves for the pre-beat glyphs and then use this width to shift over the grace notes instead of the "shared reserved space" which might be unused. 


**Edit:**
 I reconstructed your example and extended it with some variants which should make it a bit more visible how the arrangement with these glyphs is. While there is still a bit padding we could maybe eliminate in some cases, the problem you report is a very special case: 

![image](https://user-images.githubusercontent.com/674916/212482138-ba912ccd-d854-4484-9855-cd78a095f02b.png)

**Comment by Danielku15** [comment, marked as answer]
I made some optimization in https://github.com/CoderLine/alphaTab/issues/1093 
You can see the new positioning of my example here: https://github.com/CoderLine/alphaTab/blob/develop/test-data/visual-tests/special-notes/grace-notes-alignment.png
